src/lib/storage/vector_compression/simd_bp128/simd_bp128_iterator.hpp

```diff
@@ -18,77 +19,46 @@ class SimdBp128Iterator : public BaseCompressedVectorIterator<SimdBp128Iterator>
+  explicit SimdBp128Iterator(std::unique_ptr<SimdBp128Decompressor>&& decompressor, const size_t absolute_index = 0u)
+      : _decompressor{std::move(decompressor)}, _absolute_index{absolute_index} {}
 
-  SimdBp128Iterator(const SimdBp128Iterator& other);
-  SimdBp128Iterator& operator=(const SimdBp128Iterator& other);
+  explicit SimdBp128Iterator(const SimdBp128Iterator& other)
+      : _decompressor{std::make_unique<SimdBp128Decompressor>(SimdBp128Decompressor(*other._decompressor))},
+        _absolute_index{other._absolute_index} {}
+
+  explicit SimdBp128Iterator(SimdBp128Iterator&& other)
+      : _decompressor{std::move(other._decompressor)}, _absolute_index{std::move(other._absolute_index)} {}
+
+  SimdBp128Iterator& operator=(const SimdBp128Iterator& other) {
+    if (this == &other) return *this;
+
+    _decompressor = std::make_unique<SimdBp128Decompressor>(SimdBp128Decompressor(*other._decompressor));
+    _absolute_index = other._absolute_index;
+    return *this;
+  }
 
-  SimdBp128Iterator(SimdBp128Iterator&& other) = default;
   SimdBp128Iterator& operator=(SimdBp128Iterator&& other) = default;
 
   ~SimdBp128Iterator() = default;
 
  private:
   friend class boost::iterator_core_access;  // grants the boost::iterator_facade access to the private interface
 
-  void increment() {
-    ++_absolute_index;
-    ++_current_meta_block_index;
+  void increment() { ++_absolute_index; }
 
-    if (_current_meta_block_index >= Packing::meta_block_size && _absolute_index < _size) {
-      _unpack_next_meta_block();
-    }
-  }
-
-  void decrement() {
-    --_absolute_index;
+  void decrement() { --_absolute_index; }
 
-    if (_current_meta_block_index > 0) {
-      --_current_meta_block_index;
-    } else {
-      _unpack_previous_meta_block();
-    }
-  }
-
-  void advance(std::ptrdiff_t n) {
-    PerformanceWarning("Using repeated increment/decrement for random access");
-    // The easy way for now
-    if (n < 0) {
-      for (std::ptrdiff_t i = n; i < 0; ++i) {
-        decrement();
-      }
-    } else {
-      for (std::ptrdiff_t i = 0; i < n; ++i) {
-        increment();
-      }
-    }
-  }
+  void advance(std::ptrdiff_t n) { _absolute_index += n; }
 
   bool equal(const SimdBp128Iterator& other) const { return _absolute_index == other._absolute_index; }
 
   std::ptrdiff_t distance_to(const SimdBp128Iterator& other) const { return other._absolute_index - _absolute_index; }
 
-  uint32_t dereference() const { return (*_current_meta_block)[_current_meta_block_index]; }
-
- private:
-  void _unpack_next_meta_block();
-  void _unpack_previous_meta_block();
-
-  void _read_next_meta_info();
-  void _read_previous_meta_info();
-  void _unpack_block(uint8_t meta_info_index);
+  uint32_t dereference() const { return _decompressor->get(_absolute_index); }
 
  private:
-  const pmr_vector<uint128_t>* _data;
-  size_t _size;
-
-  size_t _data_index;
+  std::unique_ptr<SimdBp128Decompressor> _decompressor;
```

Comment:
Why is this a pointer, not an immediate member?

Reply:
Will change.